### PR TITLE
Introduced cluster version upgrade task

### PR DESCRIPTION
**Pull request by ahmetmircik**
ee impl: https://github.com/hazelcast/hazelcast-enterprise/pull/2640

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/instance/NodeExtension.java [hunk 1]
```diff
@@ -316,4 +316,7 @@
      * Send PhoneHome ping from OS or EE instance to PhoneHome application
      */
     void sendPhoneHome();
+
+
+    Runnable createClusterVersionChanger();
```
[the remark this answers is not in the document]
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/instance/NodeExtension.java]: done

**Comment by mmedenjak** [comment]
run-lab-run

**Comment by ahmetmircik** [comment]
run-lab-run

**Comment by ahmetmircik** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembershipManager.java [hunk 2]
```diff
@@ -1211,6 +1220,10 @@ public void run() {
                 clusterService.getClusterJoinManager().reset();
                 sendMemberListToOthers();
                 logger.info("Mastership is claimed with: " + newMembersView);
+
+                // async call
+                node.getNodeExtension().scheduleClusterVersionAutoUpgrade();
```
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembershipManager.java]: this seems superfluous here, as `updateMembers(newMembersView)` above will also invoke `scheduleClusterVersionAutoUpgrade`

Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembershipManager.java]: It was needed to handle a case after split brain when master falls to second brain but all other nodes in the first brain is upgraded. In this case, we need to call upgrade task when deciding master.

**Comment by mdogan** [comment]
Can you please remove the formatting changes? Currently they are making a great noise and it's difficult to see actual changes among them.

File: hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembershipManager.java (hunk 2 above)
Comment (mdogan) [reply hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembershipManager.java]: I think @vbekiaris is right. The `updateMembers(newMembersView)` call just above already schedules the auto-upgrade task.

File: hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterServiceImpl.java [hunk 3]
```diff
@@ -867,9 +854,13 @@ public HotRestartService getHotRestartService() {
 
     @Override
     public void changeClusterVersion(Version version) {
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterServiceImpl.java]: changes in this file starts here, others are formatting related.

File: hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembershipManager.java (hunk 2 above)
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembershipManager.java]: right. fixed.

**Comment by ahmetmircik** [comment]
@mdogan now seeing changes should be easier.

File: hazelcast/src/test/java/com/hazelcast/aggregation/ProjTest.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+import com.hazelcast.projection.Projections;
+import com.hazelcast.query.SqlPredicate;
+import com.hazelcast.test.HazelcastTestSupport;
+import org.junit.Test;
+
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.Map;
+
+public class ProjTest extends HazelcastTestSupport {
```
Comment (mdogan) [inline hazelcast/src/test/java/com/hazelcast/aggregation/ProjTest.java]: Leftover?

Comment (ahmetmircik) [reply hazelcast/src/test/java/com/hazelcast/aggregation/ProjTest.java]: removed leftover file.

**Pull request merged**